wgengine/monitor/monitor_windows.go

```diff
@@ -51,229 +33,136 @@ type messageOrError struct {
-			if m.stateChanged() {
-				m.logf("interface state changed (on poll)")
-				return pollStateChangedMessage{}, nil
-			}
-			m.mu.Lock()
-			if m.inFastPoll && time.Since(m.lastNetChange) > pollFastFor {
-				m.inFastPoll = false
-				m.pollTicker.Reset(pollIntervalSlow)
-			}
-			m.mu.Unlock()
+	if m.messagec != nil {
+		m.messagec <- messageOrError{
+			message: nil,
+			error:   errClosed,
 		}
 	}
-}
 
-func (m *winMon) stateChanged() bool {
-	st, err := interfaces.GetState()
-	if err != nil {
-		return false
+	if m.addressChangeCallback != nil {
+		if err := m.addressChangeCallback.Unregister(); err != nil {
+			m.logf("addressChangeCallback.Unregister error: %v", err)
+			// TODO: Maybe we can simply log and ignore Unregister errors because
+			// the callback will be unregistered on process exit anyway?
+			return err
+		}
+		m.addressChangeCallback = nil
 	}
-	changed := !st.Equal(m.lastState)
-	m.lastState = st
-	return changed
-}
 
-func (m *winMon) awaitIPAndRouteChanges() {
-	defer close(m.goroutineDoneCh)
-	for {
-		msg, err := m.getIPOrRouteChangeMessage()
-		if err == errClosed {
-			return
-		}
-		select {
-		case m.messagec <- messageOrError{msg, err}:
-		case <-m.ctx.Done():
-			return
+	if m.addressChangeCallback != nil {
+		if err := m.addressChangeCallback.Unregister(); err != nil {
+			m.logf("addressChangeCallback.Unregister error: %v", err)
+			// TODO: The same question about errors...
+			return err
 		}
+		m.addressChangeCallback = nil
 	}
+
+	return nil
 }
 
-func (m *winMon) getIPOrRouteChangeMessage() (message, error) {
-	if m.ctx.Err() != nil {
+func (m *winMon) Receive() (message, error) {
+	m.mu.Lock()
+
+	if m.addressChangeCallback == nil || m.routeChangeCallback == nil {
+		m.mu.Unlock()
+		m.logf("Receive() call on closed monitor")
 		return nil, errClosed
 	}
 
-	// TODO(bradfitz): locking ourselves to an OS thread here
-	// likely isn't necessary, but also can't really hurt.
-	// We'll be blocked in windows.WaitForMultipleObjects below
-	// anyway, so might as well stay on this thread during the
-	// notify calls and cancel funcs.
-	// Given the past memory corruption from misuse of these APIs,
-	// and my continued lack of understanding of Windows APIs,
-	// I'll be paranoid. But perhaps we can remove this once
-	// we understand more.
-	runtime.LockOSThread()
-	defer runtime.UnlockOSThread()
-
-	addrHandle, oaddr, cancel, err := notifyAddrChange()
+	// TODO: In the original version we haven't took the initial state, which looks as bug, but check anyway
+	prevState, err := interfaces.GetState()
 	if err != nil {
-		m.logf("notifyAddrChange: %v", err)
+		m.mu.Unlock()
+		m.logf("interfaces.GetState error: %v", err)
 		return nil, err
 	}
-	defer cancel()
 
-	routeHandle, oroute, cancel, err := notifyRouteChange()
-	if err != nil {
-		m.logf("notifyRouteChange: %v", err)
-		return nil, err
-	}
-	defer cancel()
+	m.messagec = make(chan messageOrError)
+	defer func() {
+		m.mu.Lock()
+		m.messagec = nil
```

Comment:
no need for zero this. I'd instead close a channel and set a bool via a sync.Once in Close.

Reply:
In my implementation `m.messagec != nil` means "someone listens". That's why I check for `nil` in `somethingChanged`. But OK, will try to change that in the way you're suggesting (assuming that I've understood well).